packages/core/src/compiler/Style.ts

```diff
@@ -1640,12 +1662,195 @@ const translateSubstsBlock = (
   );
 };
 
+//#region Block statics
+const emptyErrs = () => {
+  return { errors: [], warnings: [] };
+};
+
+const oneErr = (err: StyleError): StyleResults => {
+  return { errors: [err], warnings: [] };
+};
+
+const combineErrs = (e1: StyleResults, e2: StyleResults): StyleResults => {
+  return {
+    errors: e1.errors.concat(e2.errors),
+    warnings: e1.warnings.concat(e2.warnings),
+  };
+};
+
+const flatErrs = (es: StyleResults[]): StyleResults => {
+  return {
+    errors: _.flatMap(es, (e) => e.errors),
+    warnings: _.flatMap(es, (e) => e.warnings),
+  };
+};
+
+// Check that every shape name and shape property name in a shape constructor exists
+const checkGPIInfo = (selEnv: SelEnv, expr: GPIDecl): StyleResults => {
+  const styName: string = expr.shapeName.value;
+
+  let errors: StyleErrors = [];
+  let warnings: StyleWarnings = [];
+
+  const shapeNames: string[] = shapedefs.map((e: ShapeDef) => e.shapeType);
+  if (!shapeNames.includes(styName)) {
+    // Fatal error -- we cannot check the shape properties (unless you want to guess the shape)
+    return oneErr({ tag: "InvalidGPITypeError", givenType: expr.shapeName });
+  }
+
+  // `findDef` throws an error, so we find the shape name first (done above) to make sure the error can be caught
+  const shapeDef: ShapeDef = findDef(styName);
+  const givenProperties: Identifier[] = expr.properties.map((e) => e.name);
+  const expectedProperties: String[] = Object.entries(shapeDef.properties).map(
+    (e) => e[0]
+  );
+
+  for (let gp of givenProperties) {
+    // Check multiple properties, as each one is not fatal if wrong
+    if (!expectedProperties.includes(gp.value)) {
+      errors.push({ tag: "InvalidGPIPropertyError", givenProperty: gp });
+    }
+  }
+
+  return { errors, warnings };
+};
+
+// Check that every function, objective, and constraint exists (below) -- parametrically over the kind of function
+const checkFunctionName = (selEnv: SelEnv, expr: Expr): StyleResults => {
+  if (
+    !(expr.tag === "CompApp" || expr.tag === "ObjFn" || expr.tag === "ConstrFn")
+  ) {
+    throw Error("internal error: expected function");
+  }
+
+  const fnDict = FN_DICT[expr.tag];
+  if (!fnDict) {
+    throw Error("internal error: unexpected tag");
+  }
+
+  const fnNames: string[] = Object.entries(fnDict).map((e) => e[0]); // Names of built-in functions of that kind
+
+  const givenFnName: Identifier = expr.name;
+
+  if (!fnNames.includes(givenFnName.value)) {
+    const fnErrorType = FN_ERR_TYPE[expr.tag];
+    if (!fnErrorType) {
```

Comment:
Maybe you could avoid excessive checking like this by narrowing down the type of `expr` first to something like `ObjFn | ConstrFn | CompApp`. It's okay if that's not possible.